server/conn.go

```diff
@@ -1312,16 +1312,51 @@ func (cc *clientConn) writeResultset(ctx context.Context, rs ResultSet, binary b
 	return cc.flush()
 }
 
-func (cc *clientConn) writeColumnInfo(columns []*ColumnInfo, serverStatus uint16) error {
+func (cc *clientConn) writeColumnInfo(rs ResultSet, serverStatus uint16) error {
+	var err error
 	data := cc.alloc.AllocWithLen(4, 1024)
+	ps := rs.PrepareStmt()
+	storer := func(columnBytes []byte) {
+		if ps != nil {
+			destBytes := make([]byte, len(columnBytes))
+			copy(destBytes, columnBytes)
+			ps.ColumnsInfoBytes = append(ps.ColumnsInfoBytes, destBytes)
+		}
+	}
+	defer func() {
+		if ps != nil {
+			if err != nil {
+				ps.ColumnsInfoBytes = nil
+			}
+		}
+	}()
+	// try to use cached column info bytes writing packet directly
+	if ps != nil && ps.ColumnsInfoBytes != nil {
+		return cc.writeColumnInfoByBytes(data, ps.ColumnsInfoBytes, serverStatus)
+	}
+	columns := rs.Columns()
 	data = dumpLengthEncodedInt(data, uint64(len(columns)))
-	if err := cc.writePacket(data); err != nil {
+	storer(data[4:])
```

Comment:
We don't need to store the length packet, the length is the length of `ColumnsInfoBytes` so we can inline `storer` function into line 1345.